Give sharing support

sixstone-qq opened the pull request:
With different roles including Author, Editor or Owner.

`NEWS` tip for **Added**:
- Give support to sharing calendars, contacts, tasks and notes with different roles including Author, Editor or Folder Owner
- Fix security issue validate permissions to read root folders
- Fix security issue validate permissions to read and modify permission tables

It requires #249 to work with sharing calendar.

Commit:
oc: Query for fmids in behalf of owner's resource

Instead of using connected active user because the fmids are related
to the root folder (context) owner. This avoids returning back incorrect
identifiers which mostly collide with already associated URLs.

Commit:
oc: Get new change numbers using resource's owner

Instead of using the connected active user.

Although this provides no changes in the result, it could be depending
on changes from the backend so it'd better have it accurated to what
the OpenChange DB API offers.

Commit:
oc: Read FAI messages from shared folder must be possible

At least, to be able to read a shared folder using Online mode.
It requests the folder to know if the message can be read by
current active user.

Commit:
oc: Mimetise PidTagRights behaviour to PidTagMemberRights

As specified by [MS-OXCFOLD] Section 2.2.2.2.2.8 we exclude
FreeBusySimple and FreeBusyDetailed right flags.

Commit:
oc: Check FolderVisible permission to show permission list

When the active user is not the owner of the resource as specified
by [MS-OXCPERM] Section 3.2.5.1

Commit:
oc: Check FolderOwner right to modify permissions

As specified by [MS-OXCPERM] Section 3.2.5.2, the ModifyPermissions ROP
is only possible to users which have this right.

After this changeset, we check the active user can modify permission
list. This is a security fix.

Commit:
oc: API to know if a message can be deleted by a subscriber

Instead of asking general container. This gives the possibility to
perform the deletion depending on the data from the message, for instance,
the user creator.

Commit:
oc: Remove specialised PidTagAccess and PidTagAccessLevel

As the generic one from MAPIStoreMessage applies here since
we use subscriberCan* protocol to know the values from these
properties.

Commit:
oc: Check permissions on read opening a shared folder

This is a security issue that allowed a user to read the number
of messages and its subjects when it does not have any permission to read.

Now the user cannot see other's folder without asking for me to the owner.

Commit:
oc: Added custom MAPI permissions

This set is only available at sogo-openchange library scope but
they are useful to provide all permissions set from any MAPI client

Commit:
oc-notes: Implement edit own and delete own permissions

By storing the PidTagCreatorName on creation and checking when
trying to edit or delete an specific message.

Commit:
oc: Use the same permission set original from MAPIStoreDBFolder

To keep backwards compatibility.

Commit:
oc: Implement edit own and delete own in GCSFolders

And returning back PidTagCreatorName.

This is done by checking the owner of the resource if the given
permission is restricted to edit/delete own items.

This requires a52bc3b to work in calendar folders as it requires to store and retrieve
the MAPI custom permissions in the ACL.

Commit:
oc-calendar: Add edit/delete own and Folder Contact/Owner sharing perm

By storing these custom MAPI roles in the ACL.

The extension field 'X-SOGO-COMPONENT-CREATED-BY' is used to store the
event creator when it is done from Outlook. It is the same field SOGo
uses when an event is created from a shared folder in the webmail.

The creator and the organizer/owner of the event can be different and it can
be used from external sources by checking the organizer field. This matches
the specification from [MS-OXOCAL] Section 1.1 which defines the organizer
as the owner or creator of the event.

Commit:
oc-contacts: Add edit/delete own and Folder Contact/Owner sharing perm

By storing these custom MAPI roles in the ACL.

An extension field called 'X-OPENCHANGE-CREATOR' is created in the vcard
to validate the creator/owner of the contact in the shared folder.

Commit:
oc-tasks: Add edit/delete own and Folder Contact/Owner sharing perm

By storing these custom MAPI roles in the ACL. Take into account that
a task folder is shared with a calendar folder with the same name, therefore
permissions are shared and overwritten from different Outlook sections.

The extension 'X-SOGO-COMPONENT-CREATED-BY' is used to store the task
creator in both Outlook and SOGo Webmail.

The PidLidTaskOwner is not yet properly managed and we are always returning
the folder owner but to effects of sharing that extension is used by now
which matches a little more with what the user expects until we fix
the task ownership defined in [MS-OXOTASK].

Commit:
oc-notes: Point which folder class must be used while opening subfolders

It was using MAPIStoreDBFolder class instead of specialised version
MAPIStoreNotesFolder and thus the shared subfolders where set to create
messages as normal messages instead of notes.

blaxter commented on OpenChange/MAPIStoreContactsFolder.m:
```diff
@@ -95,15 +111,28 @@ - (uint32_t) exchangeRightsForRoles: (NSArray *) roles
+    rights |= RightsFolderOwner | RoleNone;
+
+  if ([roles containsObject: MAPIStoreRightFolderContact])
+    rights |= RightsFolderContact;
```
why the previous one has `| RoleNone` and this one hasn't?

sixstone-qq replied:
According to [MS-OXCPERM] Section 2.2.7, under `FolderVisible` flag row:
```
If the client sets the ReadAny flag or the FolderOwner flag, the client
MUST set this flag as well.
```

icandigitbaby commented:
:+1: Very very nice (and great documentation). Big kudos, @sixstone-qq

blaxter commented on OpenChange/MAPIStoreTasksFolder.m:
```diff
@@ -116,18 +133,64 @@ - (uint32_t) exchangeRightsForRoles: (NSArray *) roles
+    rights |= RightsDeleteOwn;
+
+  if ([roles containsObject: MAPIStoreRightFolderOwner])
+    rights |= RightsFolderOwner | RoleNone;
```
This `|RoleNone` is not needed because of lines 148-149 below

sixstone-qq replied:
I will move the Folder Visible above to avoid setting FolderVisible when only FolderContact was set as the user expects. This was nicely suggested from @blaxter

blaxter commented on OpenChange/MAPIStoreTasksFolder.m:
```diff
@@ -116,18 +133,64 @@ - (uint32_t) exchangeRightsForRoles: (NSArray *) roles
+
+  if (!modifierRoles)
+    modifierRoles = [[NSArray alloc] initWithObjects:
+                                       SOGoCalendarRole_PublicModifier,
```
2 spaces :o, but I don't mind (i just want to put something because i can't spot any bug)

sixstone-qq replied:
I do mind :dancer: Thanks very much for your review guys

blaxter commented:
I've found an issue when we set permissions from sogo side. The problem is that we only set the maximum permission (e.g. PublicModifier = write permissions) so there are interactions currently broken.

For example if you share a calendar with this permissions:
![](http://i.snag.gy/8GWxv.jpg)

If you try to add that calendar on outlook you cannot. The internal are that on `MAPIStoreCalendarFolder.m` the method `subscriberCanReadMessages` returns `NO` because `[self activeUserRoles]` has only `(ObjectEraser, ObjectCreator, ConfidentialModifier, PrivateModifier, PublicModifier, AuthorizedSubscriber)` but we are searching for `(PublicViewer, PublicDAndTViewer, PrivateViewer, PrivateDAndTViewer, ConfidentialViewer, ConfidentialDAndTViewer)`.

We could modify this method so does something like:

    return [[self activeUserRoles] firstObjectCommonWithArray: viewerRoles] != nil
            || [self subscriberCanModifyMessages];

But then I can open the calendar but I cannot view the individual events (probably for the same reason).

sixstone-qq commented:
@blaxter You're right. My suggestion is to leave as it is right now, it will work if you reset the permissions in Outlook client and implement the inclusion of lower access permissions from SOGo Web UI in other PR.

Commit:
oc: Honor calendar WebMail UI permissions on OpenChange library

By expanding roles from the given ACL to have these values as flags
inside the OpenChange library. This only applies to Calendar and
Tasks folders which stored four different access rights to three
different types of events/tasks.

As the events and tasks are stored in the same table, I have added
two new classes to manage permissions in the same way and this
avoids the code duplication called MAPIStoreCalTask(Folder|Message).

sixstone-qq commented:
@blaxter I've already fixed the problem with 6936b88, please let me know if this comes up with your expectations.

blaxter committed:
Fix prepareRightsForm, keep the highest role

userRights is sorted highest to lowest right so we have to assign
only the first right found for each type
